test/ffmpeg-vaapi/encode/encoder.py

```diff
@@ -107,19 +117,31 @@ def encode(self):
 
     vars(self).update(rcmodeu = self.rcmode.upper())
 
-    self.encoded = get_media()._test_artifact(
-      "{}.{}".format(self.gen_name(), self.get_file_ext()))
-
-    iopts = self.gen_input_opts()
-    oopts = self.gen_output_opts()
-
-    self.output = call(
-      "ffmpeg -hwaccel vaapi -vaapi_device /dev/dri/renderD128 -v verbose"
-      " {iopts} {oopts}".format(iopts = iopts, oopts = oopts))
-
-    self.check_output()
-    self.check_bitrate()
-    self.check_metrics()
+    if vars(self).get("r2r", None) is not None:
+      assert (self.r2r > 1) is True, "invalid r2r value"
+      r2r_md5 = []
+      vars(self).setdefault("metric", dict(type = "md5"))
+
+      for self.count in xrange(self.r2r):
+        self.encoded = get_media()._test_artifact(
+        "{}_{}.{}".format(self.gen_name(), self.count, self.get_file_ext()))
+
+        self.gen_output()
```

Comment:
You're effectively calling `self.gen_input_opts()` and `self.gen_output_opts()` over and over again, but the result is always the same.  Thus, please refactor so they are only called once.  Hint: the result of those functions can be used for both r2r and non-r2r case.

Reply:
Wait, my bad.  Actually, self.encoded changes in output opts on each iteration.

Reply:
We can solve this, however, if we remove the `opts.format(**vars(self))` from the gen_* functions and push this responsibility to the caller.

Reply:
format(**vars(self)) called from caller and removed in gen_* functions